src/core_plugins/kibana/public/dashboard/index.js

```diff
@@ -20,7 +21,30 @@ uiRoutes
   .when(DashboardConstants.LANDING_PAGE_PATH, {
     template: dashboardListingTemplate,
     controller: DashboardListingController,
-    controllerAs: 'listingController'
+    controllerAs: 'listingController',
+    resolve: {
+      dash: function ($route, Private, courier, kbnUrl) {
+        const savedObjectsClient = Private(SavedObjectsClientProvider);
+        const title = $route.current.params.title;
+        if (title) {
+          return savedObjectsClient.find({
+            search: `"${title}"`,
+            search_fields: 'title',
+            type: 'dashboard',
+          }).then(results => {
+            // The search isn't an exact match, lets see if we can find a single exact match to use
+            const matchingDashboards = results.savedObjects.filter(dashboard => dashboard.attributes.title === title);
+            if (matchingDashboards.length === 1) {
+              kbnUrl.redirect(createDashboardEditUrl(matchingDashboards[0].id));
+            } else {
+              kbnUrl.redirect(`${DashboardConstants.LANDING_PAGE_PATH}?filter="${title}"`);
+            }
```

Comment:
If you `return uiRoutes.WAIT_FOR_URL_CHANGE_TOKEN` at the end of your resolve function then the listing route will never show up and the router will wait for the `kbnUrl.redirect()` to take effect, which might take a tick or two.